### PR TITLE
AG-13115 Add checkboxLocation selection option

### PR DESCRIPTION
**Pull request by eliasmalik**
See https://ag-grid.atlassian.net/browse/AG-13115

### CODE REVIEW

File: packages/ag-grid-community/src/selection/selectAllFeature.ts [hunk 1]
```diff
@@ -198,35 +200,41 @@ export class SelectAllFeature extends BeanStub {
...
-            result = true;
-        } else if (typeof headerCheckboxSelection === 'function') {
-            result = headerCheckboxSelection(
-                this.gos.addGridCommonParams({
-                    column: this.column,
-                    colDef: this.column.getColDef(),
-                })
-            );
+        let newHeaderCheckbox = false;
+        if (typeof rowSelection === 'object') {
```
Comment (peterjrreynolds) [inline packages/ag-grid-community/src/selection/selectAllFeature.ts]:
```
const newHeaderCheckbox = typeof rowSelection === 'object';
if (newHeaderCheckbox) {
```
?

File: packages/ag-grid-community/src/validation/rules/colDefValidations.ts [hunk 2]
```diff
@@ -129,6 +129,15 @@ const COLUMN_DEFINITION_VALIDATIONS: Validations<ColDef | ColGroupDef> = {
         },
     },
 
+    cellRendererParams: {
+        validate: (colDef) => {
+            if ('checkbox' in colDef.cellRendererParams) {
```
Comment (peterjrreynolds) [inline packages/ag-grid-community/src/validation/rules/colDefValidations.ts]: Theoretically a user could have a custom cell renderer with params containing `checkbox`.

Probably want an additional check on `(colDef.rowGroup != null  || colDef.rowGroupIndex != null || colDef.cellRenderer === 'agGroupCellRenderer')`

File: packages/ag-grid-enterprise/src/groupColumn/rendering/groupCellRendererCtrl.ts [hunk 3]
```diff
@@ -725,27 +736,36 @@ export class GroupCellRendererCtrl extends BeanStub implements IGroupCellRendere
...
+            this.addDestroyFunc(() => this.destroyCheckbox());
         }
 
         this.comp.setCheckboxVisible(checkboxNeeded);
     }
 
+    private destroyCheckbox(): void {
+        this.cbComp && this.eCheckbox.removeChild(this.cbComp.getGui());
+        this.destroyBean(this.cbComp);
+        this.cbComp = undefined;
```
Comment (peterjrreynolds) [inline packages/ag-grid-enterprise/src/groupColumn/rendering/groupCellRendererCtrl.ts]: Can do `this.cbComp = this.destroyBean(this.cbComp)`

File: packages/ag-grid-enterprise/src/groupColumn/rendering/groupCellRendererCtrl.ts [hunk 4]
```diff
@@ -725,27 +736,36 @@ export class GroupCellRendererCtrl extends BeanStub implements IGroupCellRendere
...
                     callbackParams: this.params,
                     removeHidden: true,
                 },
             });
             this.eCheckbox.appendChild(cbSelectionComponent.getGui());
-            this.addDestroyFunc(() => {
-                this.eCheckbox.removeChild(cbSelectionComponent.getGui());
-                this.destroyBean(cbSelectionComponent);
-            });
+            this.addDestroyFunc(() => this.destroyCheckbox());
```
Comment (peterjrreynolds) [inline packages/ag-grid-enterprise/src/groupColumn/rendering/groupCellRendererCtrl.ts]: Can just get `destroy()` to call `this.destroyCheckbox()` (saves adding duplicate destroy funcs if the option is updated).

File: packages/ag-grid-community/src/selection/selectAllFeature.ts (hunk 1 above)
Comment (eliasmalik) [reply packages/ag-grid-community/src/selection/selectAllFeature.ts]: Done (feels like I wasn't getting the type narrowing when I tried this before, but it seems to be happy now)

File: packages/ag-grid-community/src/validation/rules/colDefValidations.ts (hunk 2 above)
Comment (eliasmalik) [reply packages/ag-grid-community/src/validation/rules/colDefValidations.ts]: Done

File: packages/ag-grid-enterprise/src/groupColumn/rendering/groupCellRendererCtrl.ts (hunk 3 above)
Comment (eliasmalik) [reply packages/ag-grid-enterprise/src/groupColumn/rendering/groupCellRendererCtrl.ts]: Done

File: packages/ag-grid-enterprise/src/groupColumn/rendering/groupCellRendererCtrl.ts (hunk 4 above)
Comment (eliasmalik) [reply packages/ag-grid-enterprise/src/groupColumn/rendering/groupCellRendererCtrl.ts]: Done

**Pull request merged**